File: Difficulty: Medium/Subarrays with sum K/subarrays-with-sum-k.cpp
// User function Template for C++

class Solution {
  public:
    int countSubarrays(vector<int> &arr, int k) {
        // code here
        int n = arr.size();
        vector<int> preSum(n, 0);
        preSum[0] = arr[0];
        
        for(int i=1; i<n; i++)
            preSum[i] = preSum[i-1] + arr[i];
        
        unordered_map<int, int> mp;
        int cnt = 0;
        for(int i=0; i<n; i++){
            if(preSum[i] == k)
                 cnt++;
            if(mp.find(preSum[i] - k) != mp.end())
                cnt += mp[preSum[i] - k];
            
            mp[preSum[i]]++;
        }
        return cnt;
    }
};

File: Difficulty: Medium/Subarrays with sum K/subarrays-with-sum-k.py
from collections import defaultdict

class Solution:
    def cntSubarrays(self, arr, k):
        cnt = 0
        sum_ = 0
        mp = defaultdict(int)

        for num in arr:
            sum_ += num

            if sum_ == k:
                cnt += 1

            if (sum_ - k) in mp:
                cnt += mp[sum_ - k]

            mp[sum_] += 1

        return cnt